### ISSUE TITLE
composite builds and --write-locks together cause problems with missing locks/modules

### ISSUE THREAD

**Issue by xenoterracide** [gradle/gradle#7726]
say I have a composite tree
```
toplevelapp
|____________liba
|____________libb
```

now lets say I keep `liba` and `libb` in separate git repos, but you have them composite like this for localdev. if you run `gradle --write-locks build` on 4.10.2 it will not output versions for `liba` and `libb` because it sees them as using their replacement versions. personally I would like to be able to say something like --write-locks-from-artifacts or something like that.

```
Found project 'project :ph-emr-event-pub' as substitute for module 'com.potrero.ph:ph-emr-event-pub'.
Found project 'project :ph-emr-repo' as substitute for module 'com.potrero.ph:ph-emr-repo'.
Found project 'project :ph-events' as substitute for module 'com.potrero.ph:ph-events'.
FAILURE: Build failed with an exception.
* What went wrong:
Could not determine the dependencies of task ':bootJar'.
> Could not resolve all task dependencies for configuration ':runtimeClasspath'.
   > Dependency lock state for configuration 'runtimeClasspath' is out of date:
       - Resolved 'com.potrero.ph:ph-emr-entity:0.4.0' which is not part of the lock state
       - Resolved 'com.potrero.ph:ph-db-migration:0.1.1' which is not part of the lock state
       - Resolved 'com.potrero.ph:ph-emr-event-type:0.1.5' which is not part of the lock state
```

however, if I add them to the lockstate then there is problems building locally with the repos in place.... I'm uncertain, but this may be related to [my semver plugin](https://plugins.gradle.org/plugin/com.xenoterracide.gradle.sem-ver) 

side note: this would be a different issue, but with the plugin publisher I was noticing that the metadata must not have been equal to the --write-locks output, because downstream was getting a newer plugin version

**Comment by ljacomet** [comment]
Work still needs to happen to have dependency locking and composite builds working together. See #4749

**Comment by ljacomet** [comment]
See description of solution already available in #4749 and comments there for potential follow up.
